Louis/update ops api

ldelossa opened the pull request:
This PR bumps claircore to v0.0.25, plumbs the updated Updates api into ClairV4, and addresses some usability issues with the client.

Commit:
bump to cc-v0.0.25

Signed-off-by: ldelossa <ldelossa@redhat.com>

Commit:
update handlers and server

Signed-off-by: ldelossa <ldelossa@redhat.com>

Commit:
update client

This commit makes the conditional response from the server an internal
property of the client and the client cache the results from the server.
Also introduces the plumbing for the update ops api

Signed-off-by: ldelossa <ldelossa@redhat.com>

ldelossa commented on httptransport/indexreporthandler.go:
```diff
@@ -12,18 +12,6 @@ import (
 	"github.com/quay/clair/v4/indexer"
 )
 
-// unmodified determines whether to return a conditonal response
```
not deleted, just moved to a place indicating its shared.

hdonnay commented on httptransport/client/differ.go:
```diff
@@ -122,19 +145,35 @@ func (c *HTTP) LatestUpdateOperations(ctx context.Context) (map[string]uuid.UUID
+		for u, ops := range cache.uo {
+			o := make([]driver.UpdateOperation, len(ops), len(ops))
+			copy(o, ops)
+			log.Printf("%v %v", o, ops)
```
```suggestion
```

hdonnay commented on httptransport/client/differ.go:
```diff
@@ -5,8 +5,8 @@ import (
 	"encoding/json"
 	"errors"
 	"fmt"
+	"log"
```
```suggestion
```

ldelossa replied:
yup, left over debug log :+1:

hdonnay commented on httptransport/updatediffhandler.go:
```diff
@@ -4,158 +4,72 @@ import (
+		var prev uuid.UUID
+		var err error
+		if param, ok := r.URL.Query()["prev"]; ok {
+			if len(param) != 0 {
```
I don't think this param can exist and not have any values. `param := r.URL.Query().Get("prev"); param != ""` would be simpler than the double-if.

ldelossa replied:
think you're right, the Get() is a nicer api for this.

hdonnay commented on httptransport/updateoperationhandler.go:
```diff
@@ -0,0 +1,126 @@
+	// handle conditional request. this is an optimization
+	if ref, err := h.serv.LatestUpdateOperation(ctx); err == nil {
+		var validator string
+		validator = `"` + ref.String() + `"`
```
```suggestion
		validator := `"` + ref.String() + `"`
```

hdonnay commented on httptransport/updateoperationhandler.go:
```diff
@@ -0,0 +1,126 @@
+		if len(param) != 0 {
+			latest = param[0]
+		}
+	}
```
```suggestion
	latest := r.URL.Query().Get("latest")
```

hdonnay commented on httptransport/client/differ.go:
```diff
@@ -122,19 +145,35 @@ func (c *HTTP) LatestUpdateOperations(ctx context.Context) (map[string]uuid.UUID
-		return nil, err
 	}
-	return m, nil
+	return nil, fmt.Errorf("%v: unexpected status: %s", req.URL.Path, res.Status)
```
In the above switch, the first case returns the map directly, and the second case makes a copy. This might be better expressed by adding a `Copy` method to the cache object, having the cases acquire and defer the lock, move this error return into the default case, and have this line return `cache.Copy(), nil`

ldelossa replied:
good call, missed that we store the returned value. Copy method sounds like a good idea.

ldelossa replied:
I added a `cache.Set` to remove any lock usage from the caller, as it got kinda weird that the caller needed to use the lock only sometimes.

Commit:
PR Changes

Commit addresses review concerns around uoCache usage along with various
paramter handling improvements

Signed-off-by: ldelossa <ldelossa@redhat.com>
